Use dune-configurator to support compiling on different operating systems

mnxn opened the pull request:
This PR uses dune-configurator in order to choose the correct library flags for a given system, which allows the project to be compiled on non-linux operating systems.

The library flags are taken from the examples Makefile: https://github.com/raysan5/raylib/blob/103df6c40876a19c78a3380e3a9068e8ae7dc8dd/examples/Makefile#L304

I've added flags for macOS, Windows, and BSD operating systems. I have tested and confirmed that the project compiles on the `macos` and `mingw64` systems. Unfortunately, I couldn't test it on BSD because I don't have a FreeBSD machine and I have had no luck with virtual machines.

tjammer commented:
Thank you so much for doing this, great stuff!
Maybe I should this about setting up a CI here, but for now I'll try my local machine (for Windows):

On my machine, using the latest OCaml for Windows from https://fdopen.github.io/opam-repository-mingw/, I run into a couple of problems:
- cmake can't find a c++ compiler, I have to set CXX manually (I guess this could also be done somehow from dune)
- On the cmake, which I got from cygwin (cmake 3.17.3), cmake no longer defines WIN32 for cygwin, and then glfw tries to use X11 instead of Win32. For Win32 to get picked, I have to set  CMAKE_LEGACY_CYGWIN_WIN32=1
- cmake does not find ar.exe (it tries to archive with CMAKE_AR-NOTFOUND). The last time I could only solve this by configuring cmake with -DCMAKE_AR=<absolute path to ar.exe>. but maybe there are also env variables which can be set accordingly (simply setting "AR" does not work however)

But even so, with -DCMAKE_AR set, it fails to link the shared lib, b/c it tries to link against "dl".

What kind of setup did you use in your test?

mnxn commented:
> On my machine, using the latest OCaml for Windows from https://fdopen.github.io/opam-repository-mingw/, I run into a couple of problems

This is pretty strange since it seemed to compile fine from my testing. Maybe CMake was tracking down a Windows install of gcc. I will take another look later today.

I think the Windows build need to use `x86_64-w64-mingw32-gcc` and `x86_64-w64-mingw32-ar` so they work inside of OCaml for Windows. I tried specifying this but it didn't work so I kept the CMake build as it was - which seemed to work on my computer.

How would you feel about ignoring CMake and just using the already existing Makefile in `src/Makefile`? Back when I was trying to write my own bindings to Raylib, I could build on Windows by running make with `CC=x86_64-w64-mingw32-gcc` and `AR=x86_64-w64-mingw32-ar`. The dune build rules already assume the presence of `make` on the system since they run with `-j6`.

tjammer commented:
I will check with my setup again, could as well be that my PATH was dirty.

I would prefer to only use make tbh, that's what I tried to do at first. But (at least on my PC on Linux), the Makefile build behaves strangely. If I use the raylib.a from the Makefile, the core_basic_window example uses 25% cpu instead of 3%ish with the cmake build. On raylib discord, they told me to just use cmake if it works, so that's what I did. 
I have not yet found out what causes this, but would be glad to get rid of the more complicated cmake build

mnxn commented:
I rewrote the dune rules to use make now. It turns out that the only reason CMake worked for me on Windows was because it was picking up the native Windows 10 CMake and not the one from Cygwin. Now that the build uses make and patches the Makefile on Windows, it seems to compile fine.

I dropped support non-mingw64 Windows system as it doesn't seem recommended to use them and I couldn't get it to compile.

There's also a patch to undefine `SUPPORT_HALFBUSY_WAIT_LOOP` like you suggested, so hopefully this has better performance on Linux.

I can confirm that it continues to build on `mingw64` and `macOS`. I still haven't figured out how to test it on FreeBSD.

tjammer commented on src/c/dune:
```diff
@@ -26,31 +31,78 @@
+    (copy vendor/src/raylib.dll dllraylib.dll)))))
+
+(rule
+ (targets libraylib.a dllraylib.dll)
```
I've tried to build it on FreeBSD in a VM. The shared library here needs to be dllraylib.so instead of dllraylib.dll

Apart from that, I could compile it. (I couldn't like against GL, but I think that's due to the VM)

tjammer commented:
Again, many thanks! Apart from the one line regarding the FreeBSD shared lib, we're can merge, I think :)

mnxn commented:
Fixed the build target.

Have you noticed any performance improvement now that `SUPPORT_HALFBUSY_WAIT_LOOP` is unset on Linux?

tjammer commented:
Yes, I have, now it behaves exactly like the cmake build, using only ~3% CPU.
